File: cli/backend/meteo/meteo_future.py
"""Alternative Weather Backend"""
import time
from datetime import datetime

from weather_core.backend import WeatherCondition
from weather_core.backend import WeatherData
from weather_core.backend import WindData


class MeteoFuture(WeatherData):
    def __new__(cls, *args, **kwargs):
        return super().__new__(cls)

    def __init__(self, forecast_json, aqi_json, metric, index):
        self.index = index
        min_temp = forecast_json["daily"]["temperature_2m_min"][self.index // 24]
        max_temp = forecast_json["daily"]["temperature_2m_max"][self.index // 24]
        wind = WindData(
            forecast_json["current_weather"]["windspeed"],
            int(forecast_json["current_weather"]["winddirection"]),
        )
        cloud_cover = forecast_json["hourly"]["cloudcover"][self.index]
        aqi = 0
        if (
            len(aqi_json["hourly"]["european_aqi"]) > self.index
            and aqi_json["hourly"]["european_aqi"][self.index] is not None
        ):
            aqi = aqi_json["hourly"]["european_aqi"][self.index] // 20
        super().__init__()
        # self.time = int(time.mktime(datetime.strptime(forecast_json["hourly"]["time"][self.index], '%y-%m-%dT%H:%M').timetuple()))
        self.time = int(time.mktime(datetime.now().timetuple()))
        self.temperature = forecast_json["current_weather"]["temperature"]
        self.min_temp = min_temp
        self.max_temp = max_temp
        self.wind = wind
        self.dewpoint = forecast_json["hourly"]["dewpoint_2m"][self.index]
        self.feels_like = forecast_json["hourly"]["apparent_temperature"][self.index]
        self.aqi = aqi
        self.cloud_cover = cloud_cover
        conditions = []
        if cloud_cover == 0:
            conditions.append(WeatherCondition(800))
        elif cloud_cover < 25:
            conditions.append(WeatherCondition(801))
        elif cloud_cover < 50:
            conditions.append(WeatherCondition(802))
        elif cloud_cover < 85:
            conditions.append(WeatherCondition(803))
        else:
            conditions.append(WeatherCondition(804))
        if (0 < forecast_json["hourly"]["rain"][self.index] < 0.098 and not metric) or (
            0 < forecast_json["hourly"]["rain"][self.index] < 2.5 and metric
        ):
            conditions.append(WeatherCondition(500))
        elif (forecast_json["hourly"]["rain"][self.index] < 0.39 and not metric) or (
            forecast_json["hourly"]["rain"][self.index] < 10 and metric
        ):
            conditions.append(WeatherCondition(501))
        elif (forecast_json["hourly"]["rain"][self.index] < 2 and not metric) or (
            forecast_json["hourly"]["rain"][self.index] < 50 and metric
        ):
            conditions.append(WeatherCondition(502))
        else:
            conditions.append(WeatherCondition(503))
        if forecast_json["hourly"]["snowfall"][self.index] != 0:
            conditions.append(WeatherCondition(601))
        self.conditions = conditions
        self.condition_ids = self.get_condition_ids()
        self.condition_sentence = self.get_conditions_sentence()

File: src/backend/meteo/meteo_future.rs
use crate::backend::meteo::meteo_json::{MeteoAirQualityJson, MeteoForecastJson};
use crate::backend::weather_condition::WeatherCondition;
use crate::backend::weather_data::{get_conditions_sentence, WeatherDataRS};
use crate::backend::wind_data::WindData;
use crate::now;

pub struct MeteoFuture {
    data: MeteoForecastJson,
    aqi: MeteoAirQualityJson,
    pub index: usize,
    metric: bool,
}

impl MeteoFuture {
    pub fn new(data: MeteoForecastJson, aqi: MeteoAirQualityJson, metric: bool, index: usize) -> Self {
        MeteoFuture {
            data,
            aqi,
            metric,
            index
        }
    }
}

impl WeatherDataRS for MeteoFuture {
    fn get_time(&self) -> i128 {
        now() as i128
    }

    fn get_temperature(&self) -> f32 {
        self.data.current_weather.temperature
    }

    fn get_min_temp(&self) -> f32 {
        self.data.daily.temperature_2m_min[0] // TODO: Fix
    }

    fn get_max_temp(&self) -> f32 {
        self.data.daily.temperature_2m_max[0] // TODO: Fix
    }

    fn get_wind(&self) -> WindData {
        WindData {
            speed: self.data.current_weather.windspeed as f64,
            heading: self.data.current_weather.winddirection as i16,
        }
    }

    fn get_raw_data(&self) -> String {
        serde_json::to_string_pretty(&self.data).expect("dump to string failed")
    }

    fn get_dewpoint(&self) -> f32 {
        self.data.hourly.dewpoint_2m[self.index]
    }

    fn get_feels_like(&self) -> f32 {
        self.data.hourly.apparent_temperature[self.index]
    }

    fn get_aqi(&self) -> u8 {
        self.aqi.hourly.european_aqi[self.index]
    }

    fn get_cloud_cover(&self) -> u8 {
        self.data.hourly.cloudcover[self.index]
    }

    fn get_conditions(&self) -> Vec<WeatherCondition> {
        let mut conditions: Vec<WeatherCondition> = Vec::new();
        let cloud_cover = self.get_cloud_cover();
        if cloud_cover == 0 {
            conditions.push(WeatherCondition::new(800));
        }
        else if cloud_cover < 25
        {
            conditions.push(WeatherCondition::new(801));
        }
        else if cloud_cover < 50 {
            conditions.push(WeatherCondition::new(802));
        }
        else if cloud_cover < 85 {
            conditions.push(WeatherCondition::new(803));
        }
        else {
            conditions.push(WeatherCondition::new(804));
        }
        if self.data.hourly.rain[self.index] != 0.0 {
            let rain = self.data.hourly.rain[self.index];
            let metric = self.metric;
            if (0.0 < rain && rain < 0.098 && !metric
            ) || (0.0 < rain && rain  < 2.5 && metric) {
                conditions.push(WeatherCondition::new(500));
            } else if (rain < 0.39
            && !metric) || (
                rain < 10.0 && metric
            )
            {
                conditions.push(WeatherCondition::new(501));
            }
            else if (rain < 2.0 && !metric) || (
                rain < 50.0 && metric
            )
            {
                conditions.push(WeatherCondition::new(502));
            }
            else if rain != 0.0 {
                conditions.push(WeatherCondition::new(503));
            }
        }
        if self.data.hourly.snowfall[self.index] != 0.0 {
            conditions.push(WeatherCondition::new(601));
        }
        return conditions;
    }

    fn get_condition_sentence(&self) -> String {
        get_conditions_sentence(self.get_conditions())
    }
}